Lint for using hand-writing a fold with the same behaviour as any

theotherphil opened the pull request:
https://github.com/rust-lang-nursery/rust-clippy/issues/2237

Some things that need to be addressed before this can be merged. (@Manishearth @oli-obk)

- [x] Do I need to make any changes to the lint name, documentation, description or suggestions?
- [x] See DONOTMERGE comment at the start of lint_fold_any - how can I check that the fold being called is actually the trait method on Iterator?
- [x] Decide whether I need to handle my ```TODO: don't suggest .any(|x| f(x)) if we can suggest .any(f)``` (and if so, work out how to do this) (Resolution: won't bother)
- [x] Reduce the span of the error to only include `.fold(...)` instead the caller as well (see comments from @killercup about enabling rustfix support)

~~I'll create a new issue for the sum and product cases and handle them in another PR.~~

killercup commented on clippy_lints/src/methods.rs:
```diff
@@ -653,7 +670,8 @@ impl LintPass for Pass {
             ITER_CLONED_COLLECT,
-            USELESS_ASREF
+            USELESS_ASREF,
+            FOLD_ANY
```
Nit: Does this macro support trailing commas? If so, please add one and the next diff is only one line :)

theotherphil replied:
It appears not:

```
error: unexpected end of macro invocation
   --> clippy_lints/src/methods.rs:674:21
    |
674 |             FOLD_ANY,
    |                     ^
```

theotherphil replied:
I've created an issue https://github.com/rust-lang-nursery/rust-clippy/issues/2352. I'll do this next weekend if it's still open.

killercup replied:
Oh, this is a lint defined in the compiler, so that issue is probably more useful over at <https://github.com/rust-lang/rust> :)

theotherphil replied:
Nuts. https://github.com/rust-lang/rust/issues/47428

killercup commented on clippy_lints/src/methods.rs:
```diff
@@ -1105,6 +1125,50 @@ fn lint_iter_cloned_collect(cx: &LateContext, expr: &hir::Expr, iter_args: &[hir
 }
 
+fn lint_fold_any(cx: &LateContext, expr: &hir::Expr, fold_args: &[hir::Expr]) {
+    // DONOTMERGE: What if this is just some other method called fold?
```
Not sure if applicable in this context but other parts of clippy use `match_trait_method(cx, expr, &paths::ITERATOR)`

theotherphil replied:
Thanks. I tried this (well, actually, I implemented the equivalent by hand as I missed this method), but this doesn't quite do what I want - it's not quite enough to know that the caller is an Iterator - this could be an inherent impl rather than the trait method. Maybe this is rare/ill-advised enough that it doesn't matter?

killercup commented on clippy_lints/src/methods.rs:
```diff
@@ -1105,6 +1125,50 @@ fn lint_iter_cloned_collect(cx: &LateContext, expr: &hir::Expr, iter_args: &[hir
+                expr.span,
+                // TODO: don't suggest .any(|x| f(x)) if we can suggest .any(f)
+                &format!(
+                    ".fold(false, |{f}, {s}| {f} || {r})) is more succinctly expressed as .any(|{s}| {r})",
```
Can you use `span_lint_and_sugg` here to get output like this:

```
error: This `.fold` can be more succinctly expressed as `.any`
   --> $DIR/methods.rs:390:13
    |
390 |     let _ = (0..3).fold(false, |acc, x| acc || x > 2);
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `.any(|x| x > 2)`
    |
    = note: `-D fold-any` implied by `-D warnings`
```

If you make sure the spans are right and exactly cover the `.fold(…)`, this link should work great with rustfix :)

theotherphil replied:
Sure, I'll just need to work out what it does first! (There's no doc comment at the moment)

killercup replied:
So, Rust's diagnostic messages (aka "fancy errors") allow you to add annotations to parts of the code. One special style of annotation is a "replacement suggestion". It looks just like the "help:" thing I posted above to a human, but [rustfix](https://github.com/killercup/rustfix) can parse it to automatically fix your code.

The mystical [`span_lint_and_sugg`](https://github.com/rust-lang-nursery/rust-clippy/blob/2f62d803aba284bfed6b5d5ef3a412496b9cc50c/clippy_lints/src/utils/mod.rs#L599) is a shortcut to create a new lint message and at the same time also add a suggestion. It's basically the same as `span_lint`, but you pass two parameters, `help: &str`, which is the "try: " in the example above, and `sugg: String`, which is the suggested replacement (something like `format!(".any(|{s}| {r})", …)` in this case here).

theotherphil replied:
Thanks anyway. I'll add a doc comment.

I've actually not quite got this working, as I've got a span pointing at the whole expression, including the range. What's the best way of restricting this to the `.fold(...)` part?

killercup commented on tests/ui/methods.rs:
```diff
@@ -385,6 +385,16 @@ fn iter_skip_next() {
+/// Checks implementation of the `FOLD_ANY` lint
+fn fold_any_ignore_initial_value_of_true() {
+    let _ = (0..3).fold(true, |acc, x| acc || x > 2);
+}
```
You might also want to add another test that doesn't trigger the lint because it's not using a boolean, e.g. `.fold(0, |acc, x| acc += if x > 2 { 1 } else { 0 })`

theotherphil replied:
Will do.

theotherphil commented:
@killercup I've addressed all of your comments so far, except for reducing the span to only include `.fold(...)`. Although adding the doc comment created a new issue - now I get a wall of warnings about doc rendering differing between old and new renderers!

killercup commented:
> reducing the span to only include .fold(...)

Not sure how to best do this, but I'm sure someone else can help us out :)

> Although adding the doc comment created a new issue - now I get a wall of warnings about doc rendering differing between old and new renderers!

Yeah, rustdoc is switching to the CommonMark standard soon. What's the errors telling you? I assume something about the `<pre>`? If so, try putting the code block in triple-backticks or ident it by 4 spaces. (Just the usual Markdown ways to mark code blocks.)

clarcharr commented:
I think that this should be extended to all of the similar methods on `Iterator`; see the original issue for some of my examples.

I don't mind just starting with `any` for now, but perhaps the lint should have a more generic name like `unnecessary_fold`.

theotherphil commented:
I agree that this should be extended to cover your other examples - I was planning on merging this case first and then adding the rest. I'm happy to change the name. Not sure if we want one lint covering all cases or one lint for each case.

theotherphil commented:
I've added tests locally to check that `match_trait_method`does what I want, and discovered that this lint (and presumably the other methods lints) doesn't work if you call fold via `Iterator::fold(...)`

clarcharr commented:
@oli-obk wouldn't this apply to the method-chain lints as well?

oli-obk commented:
Jup, the method chain lints don't work on universal method calls, just on actual method calls with dot notation

oli-obk commented on clippy_lints/src/methods.rs:
```diff
@@ -1105,6 +1125,55 @@ fn lint_iter_cloned_collect(cx: &LateContext, expr: &hir::Expr, iter_args: &[hir
+                cx,
+                FOLD_ANY,
+                expr.span,
+                // TODO: don't suggest .any(|x| f(x)) if we can suggest .any(f)
```
That's fine, we do have another lint for that ;)

If you really want to, you can extract the code from there into a separate function so you can call it from here and there and not duplicate any logic.

oli-obk commented on clippy_lints/src/methods.rs:
```diff
@@ -1105,6 +1125,55 @@ fn lint_iter_cloned_collect(cx: &LateContext, expr: &hir::Expr, iter_args: &[hir
+            span_lint_and_sugg(
+                cx,
+                FOLD_ANY,
+                expr.span,
```
You can probably use `fold_args[0].span.next_point().with_hi(fold_args[1].span.hi())` to produce the span you want.

theotherphil replied:
This works (with small tweaks). Thanks!

theotherphil commented:
Thanks for the tips @oli-obk - I'll finish this PR off this weekend.

Huntr0 commented on clippy_lints/src/utils/mod.rs:
```diff
@@ -1034,3 +1048,11 @@ pub fn type_size<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, ty: Ty<'tcx>) -> Option<u
+        PatKind::Ref(ref subpat, _) => get_arg_name(subpat),
+        _ => None,
+    }
+}
```
nit: missing newline

theotherphil replied:
Fixed, thanks

theotherphil commented:
Please don't merge just yet - I've decided to rename to unnecessary_fold and handle the other cases as part of this PR (I'm not going to worry about producing eta-reducible replacement suggestions for now).

theotherphil commented:
@oli-obk I think this is now in a mergeable state

killercup commented:
heads up: you'll most likely need to rebase on master to make this compile on the latest nightly

killercup commented on clippy_lints/src/methods.rs:
````diff
@@ -623,6 +623,29 @@ declare_lint! {
+/// ```rust
+/// let _ = (0..3).fold(false, |acc, x| acc || x > 2);
+/// ```
+/// This could be written as:
````
I'm pretty sure it's good style to put an empty line above and below code blocks in markdown.

theotherphil replied:
The exist lints in this file don't, for whatever that's worth

killercup commented on clippy_lints/src/methods.rs:
```diff
@@ -1105,6 +1131,93 @@ fn lint_iter_cloned_collect(cx: &LateContext, expr: &hir::Expr, iter_args: &[hir
+    }
+
+    assert!(fold_args.len() == 3,
+        "Expected fold_args to have three entries - the receiver, the initial value and the closure");
```
Should this actually panic? I imagine a `debug!` and `return;` suffices

theotherphil replied:
Well, it's a logic bug, so I would have thought it should panic. I don't feel strongly though - I can change if it you'd prefer.

killercup commented on clippy_lints/src/methods.rs:
```diff
@@ -1105,6 +1131,93 @@ fn lint_iter_cloned_collect(cx: &LateContext, expr: &hir::Expr, iter_args: &[hir
+                        ".{replacement}(|{s}| {r})",
+                        replacement = replacement_method_name,
+                        s = second_arg_ident,
+                        r = snippet(cx, right_expr.span, "EXPR")
```
trailing commas are good

here, and everywhere below (and above)

killercup commented on clippy_lints/src/utils/mod.rs:
```diff
@@ -596,6 +596,20 @@ pub fn span_lint_and_then<'a, 'tcx: 'a, T: LintContext<'tcx>, F>(
+///     |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `.any(|x| x > 2)`
+///     |
+///     = note: `-D fold-any` implied by `-D warnings`
+/// </pre>
```
replace `<pre>` with triple back ticks or 4-space indent

killercup commented on clippy_lints/src/methods.rs:
```diff
@@ -1105,6 +1131,93 @@ fn lint_iter_cloned_collect(cx: &LateContext, expr: &hir::Expr, iter_args: &[hir
+                    cx,
+                    UNNECESSARY_FOLD,
+                    fold_span,
+                    // TODO: don't suggest e.g. .any(|x| f(x)) if we can suggest .any(f)
```
Maybe open an issue for this and link it here?

Also, Clippy has a lint that will suggest using UFCS instead of closures, so a second run will likely trigger (and fix) this anyway.

theotherphil replied:
https://github.com/rust-lang-nursery/rust-clippy/issues/2371

theotherphil commented:
Do you know which of the AppVeyor jobs has any hope of succeeding? The first couple of failures appear to have nothing to do with my changes.

goodmanjonathan commented on clippy_lints/src/methods.rs:
```diff
@@ -623,6 +623,29 @@ declare_lint! {
+declare_lint! {
+    pub UNNECESSARY_FOLD,
+    Warn,
+    "using `fold` when a more succint alternative exists"
```
s/succint/succinct

goodmanjonathan commented on clippy_lints/src/methods.rs:
```diff
@@ -623,6 +623,29 @@ declare_lint! {
 }
 
+
+/// **What it does:** Checks for using `fold` when a more succint alternative exists.
```
s/succint/succinct

oli-obk commented:
Thanks!
